plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -190,6 +204,84 @@ public TableSnapshot loadSnapshot(ConnectorSession session, SchemaTableName tabl
+    {
+        try {
+            Optional<LastCheckpoint> lastCheckpoint = findCheckpoint(fileSystem, tableLocation, endVersion);
+            return TableSnapshot.load(
+                    table,
+                    lastCheckpoint,
+                    fileSystem,
+                    tableLocation,
+                    parquetReaderOptions,
+                    checkpointRowStatisticsWritingEnabled,
+                    domainCompactionThreshold,
+                    Optional.of(endVersion));
+        }
+        catch (UncheckedExecutionException e) {
+            throwIfUnchecked(e.getCause());
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Optional<LastCheckpoint> findCheckpoint(TrinoFileSystem fileSystem, String tableLocation, Long endVersion)
+    {
+        Optional<LastCheckpoint> lastCheckpoint = readLastCheckpoint(fileSystem, tableLocation);
+        if (lastCheckpoint.isPresent() && lastCheckpoint.get().version() <= endVersion) {
+            return lastCheckpoint;
+        }
+
+        // TODO Make this logic efficient
+        Optional<LastCheckpoint> latestCheckpoint = Optional.empty();
+        Location transactionDirectory = Location.of(getTransactionLogDir(tableLocation));
+        try {
+            FileIterator files = fileSystem.listFiles(transactionDirectory);
+            while (files.hasNext()) {
+                FileEntry file = files.next();
+                Optional<LastCheckpoint> checkpoint = extractCheckpointVersion(file);
+                if (checkpoint.isEmpty()) {
+                    continue;
+                }
+
+                long version = checkpoint.get().version();
+                if (version > endVersion || (latestCheckpoint.isPresent() && version < latestCheckpoint.get().version())) {
+                    continue;
+                }
+                latestCheckpoint = checkpoint;
+            }
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return latestCheckpoint;
+    }
+
+    private static Optional<LastCheckpoint> extractCheckpointVersion(FileEntry file)
+    {
+        String fileName = file.location().fileName();
+        Matcher classicCheckpoint = CLASSIC_CHECKPOINT.matcher(fileName);
+        if (classicCheckpoint.matches()) {
+            long version = Long.parseLong(classicCheckpoint.group(1));
+            return Optional.of(new LastCheckpoint(version, file.length(), Optional.empty(), Optional.empty()));
+        }
+
+        Matcher multiPartCheckpoint = MULTI_PART_CHECKPOINT.matcher(fileName);
+        if (multiPartCheckpoint.matches()) {
+            long version = Long.parseLong(multiPartCheckpoint.group(1));
+            int parts = Integer.parseInt(multiPartCheckpoint.group(3));
+            return Optional.of(new LastCheckpoint(version, file.length(), Optional.of(parts), Optional.empty()));
+        }
+
+        Matcher v2Checkpoint = V2_CHECKPOINT.matcher(fileName);
```

Comment:
can we order these matchers in the most to least probable order?

Reply:
I think the current order is fine from the perspective. Classic is the default style and multi-part was born before v2.